File: 2024/typescript/day18.ts
import { type DayModule, type SolutionFactory } from './types'
import * as utils from './utils'

type Pos = [number, number]
type Field = number
type Grid = Field[][]

type State = {
    grid: Grid
    bytes: Pos[]
}

function parseBytes(lines: string[]): Pos[] {
    return lines.map(line => utils.split_numbers(line, ',', 2) as Pos)
}

function init(bytes: Pos[], dropCount: number): State {
    const gridSize = Math.max(...bytes.map(pos => Math.max(...pos))) + 1
    const grid: Grid = Array.from({ length: gridSize }, () =>
        Array.from({ length: gridSize }, () => Infinity),
    )

    bytes.slice(0, dropCount).forEach(([x, y]) => {
        grid[y][x] = 0
    })

    return { grid, bytes }
}

function minimumSteps({ grid }: State): number {
    const size = grid.length
    grid[0][0] = 0

    const isEnd = ([x, y]: Pos) => x === size - 1 && y === size - 1

    const field = ([x, y]: Pos) => grid[y][x]

    function* nextSteps([x, y]: Pos): Generator<Pos> {
        if (x > 0) yield [x - 1, y]
        if (x < size - 1) yield [x + 1, y]
        if (y > 0) yield [x, y - 1]
        if (y < size - 1) yield [x, y + 1]
    }

    const isValid = ([x, y]: Pos, steps: number) => steps < grid[y][x]

    const stack: Pos[] = [[0, 0]]
    let current: Pos | undefined
    while ((current = stack.pop())) {
        if (isEnd(current)) {
            continue
        }

        const steps = field(current) + 1
        const next = [...nextSteps(current).filter(pos => isValid(pos, steps))]
        next.forEach(([x, y]) => {
            grid[y][x] = steps
        })
        stack.push(...next)
    }
    return grid[size - 1][size - 1]
}

function firstBlockingByte(state: State): string {
    const { bytes } = state
    let okDrops = 0
    let failDrops = bytes.length - 1
    while (failDrops - okDrops > 1) {
        const drops = Math.floor((okDrops + failDrops) / 2)
        const state = init(bytes, drops)
        const steps = minimumSteps(state)
        if (steps === Infinity) {
            failDrops = drops
        } else {
            okDrops = drops
        }
    }
    let [x, y] = bytes[okDrops]
    return `${x},${y}`
}

export const part1: Part = input => () =>
    minimumSteps(init(parseBytes(input), 1024)).toString()

export const part2: Part = input => () =>
    firstBlockingByte(init(parseBytes(input), 0))

export const day = import.meta.file.match(/day(\d+)/)![1]
export const input = await utils.readInputLines(day)
part1.solution = '292'
part2.solution = '58,44'

export const main = import.meta.main
if (main) {
    const module: Module = await import(import.meta.path)

    const testInput = utils.as_lines([
        `
5,4
4,2
4,5
3,0
2,1
6,3
2,4
1,5
0,6
3,3
2,6
5,1
1,2
5,5
2,5
6,5
1,4
0,4
6,4
1,1
6,1
1,0
0,5
1,6
2,0
`,
    ])

    await utils.tests(
        () =>
            utils.test_all(
                [
                    'Test part 1',
                    '22',
                    () =>
                        minimumSteps(
                            init(parseBytes(testInput[0]), 12),
                        ).toString(),
                ],
                [
                    'Test part 2',
                    '6,1',
                    () => firstBlockingByte(init(parseBytes(testInput[0]), 0)),
                ],
            ),
        () => utils.test_day(module),
    )
}

type In = typeof input
type Out = string
type Module = DayModule<Out, In>
type Part = SolutionFactory<Out, In>

File: 2024/zig/src/day18.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const aoc = @import("aoc_utils");

const Grid = aoc.Grid(Field, u8);
const Field = enum(u2) { accessible, corrupted };

const dirs = std.enums.values(aoc.Orientation);

const Variant = union(enum) {
    part1: struct { size: usize, len: usize },
    part2: struct { size: usize },
};
const Result = union(enum) {
    part1: Day.Result1,
    part2: Day.Result2,
};

pub fn solve(comptime variant: Variant, input: []const u8, gpa: Allocator) !Result {
    const size = switch (variant) {
        .part1 => |v| v.size,
        .part2 => |v| v.size,
    };
    const len = switch (variant) {
        .part1 => |v| v.len,
        .part2 => input.len,
    };

    var grid: Grid = try .initSize(size, size, .accessible, gpa);
    defer grid.deinit(gpa);

    var lines = aoc.splitLines(input);
    var bytes: std.ArrayList([]const u8) = .empty;
    defer bytes.deinit(gpa);
    var i: usize = 0;
    while (lines.next()) |line| {
        switch (variant) {
            .part1 => {
                if (i == len) {
                    try bytes.append(gpa, "0,0");
                    break;
                }
                i += 1;
            },
            .part2 => try bytes.append(gpa, line),
        }

        var it = aoc.split(line, ',');
        const x = try std.fmt.parseUnsigned(u8, it.next().?, 10);
        const y = try std.fmt.parseUnsigned(u8, it.next().?, 10);
        grid.setAt(.{ .x = x, .y = y }, .corrupted);
    }

    var stepsCount: aoc.Grid(u16, u8) = try .initSize(size, size, 0, gpa);
    defer stepsCount.deinit(gpa);

    const State = struct { pos: Grid.Pos, steps: u16 };
    var stack: std.ArrayList(State) = .empty;
    defer stack.deinit(gpa);

    var starts: std.ArrayList(State) = .empty;
    defer starts.deinit(gpa);
    try starts.append(gpa, .{ .pos = .{ .x = 0, .y = 0 }, .steps = 0 });

    while (bytes.pop()) |line| {
        var it = aoc.split(line, ',');
        const x = try std.fmt.parseUnsigned(u8, it.next().?, 10);
        const y = try std.fmt.parseUnsigned(u8, it.next().?, 10);
        grid.setAt(.{ .x = x, .y = y }, .accessible);

        try stack.appendSlice(gpa, starts.items);
        starts.clearRetainingCapacity();
        while (stack.pop()) |current| {
            switch (variant) {
                .part1 => {},
                .part2 => {
                    if (current.pos.x == size - 1 and current.pos.y == size - 1) {
                        return .{ .part2 = line };
                    }

                    if (grid.at(current.pos) == .corrupted) {
                        try starts.append(gpa, current);
                        continue;
                    }
                },
            }

            const steps = stepsCount.at(current.pos);
            if (steps != 0 and steps <= current.steps) continue;

            stepsCount.setAt(current.pos, current.steps);
            const nextSteps = current.steps + 1;
            for (dirs) |dir| {
                if (dir.nextOrNull(grid, current.pos)) |pos| {
                    switch (variant) {
                        .part1 => if (grid.at(pos) == .accessible) {
                            try stack.append(gpa, .{ .pos = pos, .steps = nextSteps });
                        },
                        .part2 => {
                            const array = switch (grid.at(pos)) {
                                .corrupted => &starts,
                                .accessible => &stack,
                            };
                            try array.append(gpa, .{ .pos = pos, .steps = nextSteps });
                        },
                    }
                }
            }
        }

        switch (variant) {
            .part1 => {
                return .{ .part1 = stepsCount.at(.{ .x = size - 1, .y = size - 1 }) };
            },
            .part2 => {},
        }
    }

    unreachable;
}

pub fn part1(input: []const u8, gpa: Allocator) !Day.Result1 {
    const result = try solve(.{ .part1 = .{ .size = 71, .len = 1024 } }, input, gpa);
    return result.part1;
}

pub fn part2(input: []const u8, gpa: Allocator) !Day.Result2 {
    const result = try solve(.{ .part2 = .{ .size = 71 } }, input, gpa);
    return result.part2;
}

pub const Day = aoc.DayInfo("18", u32, []const u8, 292, "58,44", @This(), &.{.{
    .expected1 = null,
    .expected2 = null,
    .input =
    \\
    ,
}});

test "samples 1 solve 1 + 2" {
    const input =
        \\5,4
        \\4,2
        \\4,5
        \\3,0
        \\2,1
        \\6,3
        \\2,4
        \\1,5
        \\0,6
        \\3,3
        \\2,6
        \\5,1
        \\1,2
        \\5,5
        \\2,5
        \\6,5
        \\1,4
        \\0,4
        \\6,4
        \\1,1
        \\6,1
        \\1,0
        \\0,5
        \\1,6
        \\2,0
    ;

    const expected1: Day.Result1 = 22;
    const result1 = try solve(.{ .part1 = .{ .size = 7, .len = 12 } }, input, std.testing.allocator);
    try std.testing.expectEqual(expected1, result1.part1);

    const expected2: Day.Result2 = "6,1";
    const result2 = try solve(.{ .part2 = .{ .size = 7 } }, input, std.testing.allocator);
    try std.testing.expectEqualStrings(expected2, result2.part2);
}

test "samples 1" {
    try Day.testPart1Samples();
}
test "samples 2" {
    try Day.testPart2Samples();
}
test "part 1" {
    try Day.testPart1();
}
test "part 2" {
    try Day.testPart2();
}
